overlord/devicemgmtstate/devicemgmtmgr.go

```diff
@@ -0,0 +1,561 @@
+		validate.Set("id", msg.ID())
+		validate.WaitFor(t)
+		chg.AddTask(validate)
+	}
+
+	if len(ms.PendingMessages) > 0 {
+		m.state.EnsureBefore(0)
+	}
+
+	return nil
+}
+
+// getMessageHandler retrieves a pending message and its corresponding handler.
+// Caller must hold state lock.
+func (m *DeviceMgmtManager) getMessageAndHandler(t *state.Task, ms *DeviceMgmtState) (*PendingMessage, MessageHandler, error) {
+	var id string
+	err := t.Get("id", &id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	msg, ok := ms.PendingMessages[id]
+	if !ok {
+		return nil, nil, fmt.Errorf("message %s not found in pending messages", id)
+	}
+
+	handler, ok := m.handlers[msg.Kind]
+	if !ok {
+		return nil, nil, fmt.Errorf("no handler registered for message kind %q", msg.Kind)
+	}
+
+	return msg, handler, nil
+}
+
+// doValidateMessage performs snapd-level and subsystem-level validation on a message.
+// Creates apply-message task if validation succeeds, or queue-response for rejected messages.
+// TODO: implement device targeting check
+// TODO: implement time constraint checks
+// TODO: implement assumes validation
+func (m *DeviceMgmtManager) doValidateMessage(t *state.Task, _ *tomb.Tomb) error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	msg, handler, err := m.getMessageAndHandler(t, ms)
+	if err != nil {
+		return err
+	}
+
+	var next *state.Task
+	err = handler.Validate(m.state, msg)
+	if err != nil {
+		next = m.state.NewTask("queue-response", fmt.Sprintf("Queue response for message %s", msg.ID()))
+		next.Set("status", asserts.MessageStatusRejected)
+		next.Set("body", map[string]any{"message": err.Error()})
+	} else {
+		next = m.state.NewTask("apply-message", fmt.Sprintf("Apply message %s", msg.ID()))
+	}
+
+	next.Set("id", msg.ID())
+	next.WaitFor(t)
+	t.Change().AddTask(next)
+
+	m.setState(ms)
+
+	return nil
+}
+
+// doApplyMessage dispatches the message to its subsystem handler for processsing
+// and creates an await-subsystem-change task to monitor completion.
+func (m *DeviceMgmtManager) doApplyMessage(t *state.Task, _ *tomb.Tomb) error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	ms, err := m.getState()
+	if err != nil {
+		return err
+	}
+
+	msg, handler, err := m.getMessageAndHandler(t, ms)
+	if err != nil {
+		return err
+	}
+
+	subSysChangeID, err := handler.Apply(m.state, msg)
+	if err != nil {
+		return fmt.Errorf("cannot apply message: %w", err)
+	}
+	msg.ChangeID = subSysChangeID
+
+	await := m.state.NewTask("await-subsystem-change", fmt.Sprintf("Await %s subsystem change for message %s", msg.Kind, msg.ID()))
+	await.Set("id", msg.ID())
+	await.WaitFor(t)
+	t.Change().AddTask(await)
+
+	m.setState(ms)
+
+	return nil
+}
+
+// doAwaitSubsystemChange monitors a subsystem change until completion and creates
+// a queue-response task with the processing results.
+func (m *DeviceMgmtManager) doAwaitSubsystemChange(t *state.Task, _ *tomb.Tomb) error {
```

Comment:
This doesn't need to be its own task, I think. The queue-response can have an early check for the confdb change and retry